pkg/cmd/run.go

```diff
@@ -131,52 +133,97 @@ doppler run --command "YOUR_COMMAND && YOUR_OTHER_COMMAND"`,
 			exitOnWriteFailure: exitOnWriteFailure,
 			passphrase:         passphrase,
 		}
-		secrets := fetchSecrets(localConfig, enableCache, fallbackOpts, metadataPath, nameTransformer, dynamicSecretsTTL)
+		// retrieve secrets
+		dopplerSecrets := fetchSecrets(localConfig, enableCache, fallbackOpts, metadataPath, nameTransformer, dynamicSecretsTTL)
 
 		if preserveEnv {
 			utils.LogWarning("Ignoring Doppler secrets already defined in the environment due to --preserve-env flag")
 		}
 
-		env := os.Environ()
+		mountPath := cmd.Flag("mount").Value.String()
+		mountFormat := cmd.Flag("mount-format").Value.String()
+		maxReads := utils.GetIntFlag(cmd, "mount-max-reads", 32)
+		// only auto-detect the format if it hasn't been explicitly specified
+		shouldAutoDetectFormat := !cmd.Flags().Changed("mount-format")
+
+		originalEnv := os.Environ()
 		existingEnvKeys := map[string]bool{}
-		for _, envVar := range env {
+		for _, envVar := range originalEnv {
 			// key=value format
 			parts := strings.SplitN(envVar, "=", 2)
 			key := parts[0]
 			existingEnvKeys[key] = true
 		}
 
+		secrets := map[string]string{}
 		excludedKeys := []string{"PATH", "PS1", "HOME"}
-		for name, value := range secrets {
+		for name, value := range dopplerSecrets {
 			useSecret := true
 			for _, excludedKey := range excludedKeys {
 				if excludedKey == name {
+					utils.LogDebug(fmt.Sprintf("Ignoring restricted secret %s", name))
 					useSecret = false
 					break
 				}
 			}
+			if !useSecret {
+				continue
+			}
 
-			if useSecret && preserveEnv {
+			if preserveEnv && existingEnvKeys[name] == true {
```

Comment:
Should we explicitly disallow the use of `--preserve-env` with `--mount`? Or does it still make sense to make that option available?

As it stands, there's some weird behavior where if you have a secret called `MY_SECRET` and you do `MY_SECRET=123 doppler run --mount secrets.json --preserve-env -- cat secrets.json`, you'll get a `secrets.json` which is missing `MY_SECRET`.

Reply:
Added a test for this!